### PR TITLE
Refactor uploads 

### LINKED ISSUE #2886: Refactor UploadActivity
**Summary:** 
The current flow of UploadActivity has a lot of code in a single place, because of which it is very difficult to find bugs and is very difficult to maintain the code. As a part of adopting the MVP pattern, refractor the Upload Activity to use MVP and switch to fragments instead of the entire code being in the single Activity forcing the Activity to juggle views (inside the ViewFlipper, adding complexity). Also solve #2832

**Would you like to work on the issue?**
Yes

### PR DESCRIPTION
**Pull request by ashishkumar468**
Refactor Upload Activities
Optimised UploadActivity, added fragments for various stages of upload, switched to MVP
Fixes #2886
To be tested
Uploads:

- [x] Camera upload

- [x] Camera nearby upload with successful wikidata edit

- [x] Single gallery upload

- [x] Google Photos upload

- [x] Multiple gallery upload

- [x] Gallery upload with successful wikidata edit **- upload is successful but Wikidata edit is not. The toast is not even triggered.**

- [x] Image taken date is correct

- [x] Suffixes are added if same title is used

In the current ShareActivity:

- [x] Tooltips for title, desc, license fields - **tooltip not present for title**

- [x] Language selector for desc

- [x] Prefilling of title and desc when upload initiated via Nearby

- [x] License selector

- [x] Link to license and Commons policies - **works, but crashes app**

- [x] Zoom in (and back out)

- [x] Open location of image on maps

- [x] Copy previous title/desc

Category suggestions/selection:

- [x] When search text field is empty, display:-Recent categories - **does not work**

- [x] Location-based category suggestions (if image has geolocation) - **does not work**

- [x] Title-based category suggestions - **does not work**

- [x] Wikidata item category (if image is uploaded via Nearby and if that Wikidata item has a category) - **does not work**

- [x] When user types in search text field:

Check for categories shortly after typing stops, repeatedly

- [x] Display exact matches at the top

- [x] Categories selected via both of these suggestion types need to be added to the list of category selections

- [x] **BUG:** Categories are not saved with multiple uploads, the final product in Commons is missing categories

Warnings:

- [x] If image is too dark

- [x] If image already exists in Commons (via hash check)

- [x] If user tries to submit empty title (do not allow to proceed)

- [x] If user does not select any categories

Image templates:

- [x] Ensure the same templates are added (not sure if this will be an issue if backend has not changed, but make sure you check)

### CODE REVIEW

**Comment by ashishkumar468** [comment]
Attaching screenshot for dark image warning
device-2019-06-02-214236

**Comment by ashishkumar468** [comment]
@maskaravivek Can you ebaborate this one Ensure the same templates are added (not sure if this will be an issue if backend has not changed, but make sure you check)

**Comment by vanshikaarora** [comment]
@ashishkumar468  I was checking for uploads for images. And currently I think we couldn't merge this branch into master. I have mostly focused on changing title to captions mostly the UI part. The image title for uploading the image is not the captions nor the filename. I'll take the task of First Caption as filename and sending through API. Till then we can focus on other Issues with this PR:)

**Comment by misaochan** [comment]
Just spoke to @ashishkumar468 and this is ready for testing now. I will start tomorrow - as it is a HUGE change, help with testing is always appreciated. :)

File: app/src/main/java/fr/free/nrw/commons/upload/ThumbnailsAdapter.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+import butterknife.BindView;
+import butterknife.ButterKnife;
+import com.facebook.drawee.view.SimpleDraweeView;
+import fr.free.nrw.commons.R;
+import fr.free.nrw.commons.filepicker.UploadableFile;
+import java.io.File;
+import java.util.ArrayList;
+import java.util.List;
+
```
Comment (misaochan) [inline app/src/main/java/fr/free/nrw/commons/upload/ThumbnailsAdapter.java]: Could we get Javadocs for this class and its methods please?

File: app/src/main/java/fr/free/nrw/commons/upload/UploadModel.java [hunk 2]
```diff
@@ -110,51 +150,18 @@ private UploadItem getUploadItem(UploadableFile uploadableFile,
...
-        uploadItem.first = true;
-
-        if (place != null) {
-            uploadItem.title.setTitleText(place.getName());
-            uploadItem.descriptions.get(0).setDescriptionText(place.getLongDescription().equals("?")?"":place.getLongDescription());
-            //TODO figure out if default descriptions in other languages exist
+        UploadItem uploadItem = new UploadItem(uploadableFile.getContentUri(), Uri.parse(uploadableFile.getFilePath()),
+                uploadableFile.getMimeType(context), source, gpsExtractor, place, fileCreatedDate,
+                createdTimestampSource);
+        if(place!=null){
```
Comment (misaochan) [inline app/src/main/java/fr/free/nrw/commons/upload/UploadModel.java]: Please clean up formatting to match our current code style :)

File: app/src/main/java/fr/free/nrw/commons/repository/UploadRemoteDataSource.java [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+import io.reactivex.Single;
+
+import java.util.Comparator;
+import java.util.List;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+@Singleton
+public class UploadRemoteDataSource {
```
Comment (misaochan) [inline app/src/main/java/fr/free/nrw/commons/repository/UploadRemoteDataSource.java]: Thank you for the method Javadocs. The class itself needs a Javadoc to summarize what it does though, especially if it isn't a regular Fragment/Activity/etc.

File: app/src/main/java/fr/free/nrw/commons/repository/UploadRepository.java [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+import io.reactivex.Single;
+
+import java.util.Comparator;
+import java.util.List;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+@Singleton
+public class UploadRepository {
```
Comment (misaochan) [inline app/src/main/java/fr/free/nrw/commons/repository/UploadRepository.java]: Same with this class

File: app/src/main/java/fr/free/nrw/commons/upload/categories/CategoriesPresenter.java [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+import java.util.List;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+
+import timber.log.Timber;
+
+@Singleton
+public class CategoriesPresenter implements CategoriesContract.UserActionListener {
```
Comment (misaochan) [inline app/src/main/java/fr/free/nrw/commons/upload/categories/CategoriesPresenter.java]: Javadocs please

File: app/src/main/java/fr/free/nrw/commons/upload/license/MediaLicensePresenter.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+import fr.free.nrw.commons.upload.license.MediaLicenseContract.View;
+
+import java.lang.reflect.Proxy;
+import java.util.List;
+
+import javax.inject.Inject;
+
+import timber.log.Timber;
+
+public class MediaLicensePresenter implements MediaLicenseContract.UserActionListener {
```
Comment (misaochan) [inline app/src/main/java/fr/free/nrw/commons/upload/license/MediaLicensePresenter.java]: Javadocs please

**Comment by ashishkumar468** [comment]
Thanks for testing this out @misaochan.  I will make the necessary changes and let you know :-)

**Comment by maskaravivek** [comment]
Just tested wikidata edits. They are working fine for me.

**Comment by maskaravivek** [comment]
Just checked the following: 

- title based suggestions
- tooltip for title 

These things are now working for me. 

There's still a memory leak happening which needs to be fixed (This is different from what I fixed in #3001 ). 

device-2019-06-10-111428

**Comment by ashishkumar468** [comment]
@maskaravivek  Thanks for pointing out the memory leak. I have added a fix it as well. @misaochan  @neslihanturan, The PR is up for testing :-)

**Comment by neslihanturan** [comment]
Hi @ashishkumar0207 , I get this error on your branch just after choosing and image from gallery. I am not sure if this is relevant with your changes though. On beta cluster, API 26 emulator.

```
 Caused by: java.lang.IllegalStateException: The application's PagerAdapter changed the adapter's contents without calling PagerAdapter#notifyDataSetChanged! Expected adapter item count: 0, found: 3 Pager id: fr.free.nrw.commons.beta:id/vp_upload Pager class: class fr.free.nrw.commons.contributions.UnswipableViewPager Problematic adapter: class fr.free.nrw.commons.upload.UploadActivity$UploadImageAdapter
        at androidx.viewpager.widget.ViewPager.populate(ViewPager.java:1143)
        at androidx.viewpager.widget.ViewPager.populate(ViewPager.java:1092)
        at androidx.viewpager.widget.ViewPager.setOffscreenPageLimit(ViewPager.java:856)
        at fr.free.nrw.commons.upload.UploadActivity$UploadImageAdapter.getCount(UploadActivity.java:434)
        at androidx.viewpager.widget.ViewPager.dataSetChanged(ViewPager.java:1023)
        at androidx.viewpager.widget.ViewPager$PagerObserver.onChanged(ViewPager.java:3097)
        at androidx.viewpager.widget.PagerAdapter.notifyDataSetChanged(PagerAdapter.java:291)
        at fr.free.nrw.commons.upload.UploadActivity$UploadImageAdapter.setFragments(UploadActivity.java:425)
        at fr.free.nrw.commons.upload.UploadActivity.receiveSharedItems(UploadActivity.java:343)
        at fr.free.nrw.commons.upload.UploadActivity.lambda$uFmJNhak7opat5jgMvaezcNPkNM(Unknown Source:0)
        at fr.free.nrw.commons.upload.-$$Lambda$UploadActivity$uFmJNhak7opat5jgMvaezcNPkNM.run(Unknown Source:2)
        at fr.free.nrw.commons.utils.PermissionUtils$1.onPermissionGranted(PermissionUtils.java:114)
        at com.karumi.dexter.MultiplePermissionsListenerToPermissionListenerAdapter.onPermissionsChecked(Unknown Source:35)
```

**Comment by ashishkumar468** [comment]
@neslihanturan Thanks for helping with the logs, I have raise a PR fixing the Issue. please have a look https://github.com/commons-app/apps-android-commons/pull/3012

**Comment by neslihanturan** [comment]
Closing this one since same changes are merged on another pull request.

**Pull request merged**